### DISCUSSION TITLE
QPainter::begin: Paint device returned engine == 0, type: 1

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [pyvista/pyvistaqt#194] [Q&A]
When using `QtInteractor` (Windows 10 with PySide6)

```python
from pyvistaqt import MainWindow, QtInteractor
from qtpy import QtWidgets

class Window(MainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle('Example')

        self.layout_ = QtWidgets.QVBoxLayout()

        self.container = QtWidgets.QFrame()
        self.container.setLayout(self.layout_)
        self.setCentralWidget(self.container)

        self.plotter = QtInteractor(parent=self.container)
        self.layout_.addWidget(self.plotter)
        self.plotter.show()

app = QtWidgets.QApplication([])
window = Window()
window.show()
app.exec_()
```

I'm seeing the error:

```
QPainter::begin: Paint device returned engine == 0, type: 1
```

I know `PyQt6` and `PySide6` use "QOpenGLWidget" and the others "QGLWidget" in `rwi.py`.

However, I also know [PyQt's sip supports cooperative multiple inheritance](https://www.riverbankcomputing.com/static/Docs/PyQt5/multiinheritance.html#support-for-cooperative-multi-inheritance), but [PySide's Shiboken is still struggling](https://bugreports.qt.io/browse/PYSIDE-1564). We first encountered this [here](https://github.com/pyvista/pyvistaqt/pull/12).

If I print out the `__mro__` of `QtInteractor`, I get

```
(<class 'pyvistaqt.plotting.QtInteractor'>, <class 'pyvistaqt.rwi.QVTKRenderWindowInteractor'>, <class 'PySide6.QtOpenGLWidgets.QOpenGLWidget'>, <class 'PySide6.QtWidgets.QWidget'>, <class 'PySide6.QtCore.QObject'>, <class 'PySide6.QtGui.QPaintDevice'>, <class 'Shiboken.Object'>, <class 'pyvista.plotting.plotting.BasePlotter'>, <class 'pyvista.plotting.picking.PickingHelper'>, <class 'pyvista.plotting.widgets.WidgetHelper'>, <class 'object'>)
```

Does anyone know if this is a `QOpenGLWidget` vs `QGLWidget` issue, or a `PyQt` vs `PySide` issue? Has anyone else come across this?

**UPDATE:**

When I close the application, I get a bunch of the following repeated errors:

```
2022-07-08 18:42:04.838 (  23.428s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (00000260D8C8ABD0): wglMakeCurrent failed in MakeCurrent(), error: ☠�ɠ
```

From the [PySide6 docs](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#more):

> Your widget’s OpenGL rendering context is made current when [paintGL()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.paintGL) , [resizeGL()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.resizeGL) , or [initializeGL()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.initializeGL) is called. If you need to call the standard OpenGL API functions from other places (e.g. in your widget’s constructor or in your own paint functions), you must call [makeCurrent()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.makeCurrent) first.
> 
> All rendering happens into an OpenGL framebuffer object. [makeCurrent()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.makeCurrent) ensure that it is bound in the context. Keep this in mind when creating and binding additional framebuffer objects in the rendering code in [paintGL()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.paintGL) . Never re-bind the framebuffer with ID 0. Instead, call [defaultFramebufferObject()](https://doc.qt.io/qtforpython/PySide6/QtOpenGLWidgets/QOpenGLWidget.html#PySide6.QtOpenGLWidgets.PySide6.QtOpenGLWidgets.QOpenGLWidget.defaultFramebufferObject) to get the ID that should be bound.

So I wonder if `Shiboken.Object` is breaking the inheritance chain during `super()` calls to `__init__`?

**UPDATE 2:**

```python
Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvista
>>> pyvista.Report(gpu=False)

--------------------------------------------------------------------------------
  Date: Fri Jul 08 19:18:53 2022 Pacific Daylight Time

                OS : Windows
            CPU(s) : 96
           Machine : AMD64
      Architecture : 64bit
               RAM : 190.7 GiB
       Environment : Python
       GPU Details : None

  Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64
  bit (AMD64)]

           pyvista : 0.35.1
               vtk : 9.1.0
             numpy : 1.23.0
           imageio : 2.19.3
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.5.2
         pyvistaqt : 0.9.0
           IPython : 8.4.0
              tqdm : 4.64.0
            meshio : 5.3.4
--------------------------------------------------------------------------------
>>> pyvista.Report(gpu=True)  

--------------------------------------------------------------------------------
  Date: Fri Jul 08 19:19:21 2022 Pacific Daylight Time

                OS : Windows
            CPU(s) : 96
           Machine : AMD64
      Architecture : 64bit
               RAM : 190.7 GiB
       Environment : Python
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : Quadro RTX 8000/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 512.78

  Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64
  bit (AMD64)]

           pyvista : 0.35.1
               vtk : 9.1.0
             numpy : 1.23.0
           imageio : 2.19.3
           appdirs : 1.4.4
            scooby : 0.5.12
        matplotlib : 3.5.2
         pyvistaqt : 0.9.0
           IPython : 8.4.0
              tqdm : 4.64.0
            meshio : 5.3.4
--------------------------------------------------------------------------------
>>>
```

**Comment by larsoner** [comment]
Can you see if you experience the same problem when using BackgroundPlotter? If not then you might be able to look at the BackgroundPlotter code to see how it's avoiding the problem.

My guess is that your call to `plotter.show()` before you've done `mainwindow.show()` is a problem

**Reply by adam-grant-hendry** [reply]
I get the same error if I use `BackgroundPlotter` or if I switch to show the `plotter` after the `mainwindow.show()`

**Reply by adam-grant-hendry** [reply]
@larsoner So there are 2 errors occurring. The first is

```
QPainter::begin: Paint device returned engine == 0, type: 1
```

issuing when the window is made and on every `update`.

The second is

```
2022-07-08 18:42:04.838 (  23.428s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (00000260D8C8ABD0): wglMakeCurrent failed in MakeCurrent(), error: ☠�ɠ
```

#### 1st Error

From the [PySide6 QPaintEngine docs](https://doc.qt.io/qtforpython-6/PySide6/QtGui/QPaintEngine.html#PySide6.QtGui.PySide6.QtGui.QPaintEngine.Type), this error is saying that `Qt` thinks I am on `X11` (i.e. Linux; constant `QPaintEngine.X11 == 0`), but I am really on Windows (constant `QPaintEngine.Windows == 1`)

There is a widget attribute set in `pyvistaqt.rwi::QVTKRenderWindowInteractor.__init__()`:

```python
self.setAttribute(WidgetAttribute.WA_PaintOnScreen)
```

The [PySide6.QtCore.Qt docs](https://doc.qt.io/qtforpython/PySide6/QtCore/Qt.html) state:

```
**Note:** This flag is only supported on X11 and it disables double buffering.
```

I believe this is the source of the error, so [QVTKRenderWindowInteractor.py](https://gitlab.kitware.com/vtk/vtk/-/blob/master/Wrapping/Python/vtkmodules/qt/QVTKRenderWindowInteractor.py) will need to be modified accordingly.

There is a posted `Qt` bug that may interest you and the developers regarding offscreen rendering:

- [Hidden QOpenGLWidget does not render()](https://bugreports.qt.io/browse/QTBUG-47185)
- [Easiest way for offscreen rendering with QOpenGLWidget](https://stackoverflow.com/questions/31323749/easiest-way-for-offscreen-rendering-with-qopenglwidget)
#### 2nd Error

I found I get the second error because `plotter` isn't getting closed when `MainWindow` is . (`BackgroundPlotter` does not have this problem.)

To fix this error with `QtInteractor`, simply connect the `pyvistaqt.MainWindow.signal_close` signal to `plotter.close`:

```python
from pyvistaqt import MainWindow, QtInteractor
from qtpy import QtWidgets

class Window(MainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle('Example')

        self.layout_ = QtWidgets.QVBoxLayout()

        self.container = QtWidgets.QFrame()
        self.container.setLayout(self.layout_)
        self.setCentralWidget(self.container)

        self.plotter = QtInteractor(parent=self.container)
        self.layout_.addWidget(self.plotter)
        self.plotter.show()

        self.signal_close.connect(self.plotter.close)  # Fixes 2nd problem


app = QtWidgets.QApplication([])
window = Window()
window.show()
app.exec_()
```

This fix is consistent with several other posts I found on the same topic:

- [wglMakeCurrent failed in MakeCurrent after closed a window with two vtk widget](https://discourse.vtk.org/t/wglmakecurrent-failed-in-makecurrent-after-closed-a-window-with-two-vtk-widget/5899)
- [Qt Vtk wglMakeCurrent failed in MakeCurrent(), error: invalid handle](https://www.pudn.com/news/6228d5529ddf223e1ad1dbe6.html)

This may be purely a Window issue (as I suspect since `vtkWin32OpenGLRenderWin` issues the error). There's a now [dead link](http://www.archivum.info/qt-interest@trolltech.com/2006-11/00187/Re-Multiple-GL-contexts-in-Qt-4.2-not-working.html) that claimed this was due to the fact that

> if the thread calling `makeCurrent()` is not equal to the thread that retrieved the device context, `GetDC()` is called. As stated in the linked thread, the problem was that `ReleaseDC()` was not called accordingly, causing the handle to leak and at some point triggering Windows to return `NULL` on the call to `GetDC()`, which made wglMakeCurrent() fail.

I believe the same conclusion is alluded to [here](https://markmail.org/message/w24u2o26zsebss65#query:+page:1+mid:s5zjzadisvcpwckz+state:results) and [here](https://vtkusers.public.kitware.narkive.com/InX9JnDR/error-during-cleanup-wglmakecurrent-failed-in-clean).

**Reply by adam-grant-hendry** [reply]
@larsoner This appeared almost 2 years ago in a related issue on mayavi [Problem when using QGLWidget](https://github.com/enthought/mayavi/issues/969#)

After debugging in VSCode with the following `launch.json` so it will go into packages:

```json
{
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "justMyCode": false
        }
    ],
}
```

the culprit is `rwi.py::QVTKRenderWindowInteractor.paintEngine()` getting called when a `Paint` event occurs:

```python
def paintEngine(self):
    return None
```

My guess is `paintEngine` was overriden to return `None` (NULL in C++) to avoid the error:

```
QWidget::paintEngine: Should no longer be called
```
